### PR TITLE
Fixes #1514, #321. Fixes -SNAPSHOT issue by re-implemeting ChainResolver

### LINKED ISSUE #1514: sbt doesn't see new SNAPSHOT dependency versions
This is a general case for #321. 
## steps
1. Create two repositories RepoA and RepoB (or use `Resolver.defaultShared.nonlocal()`)
2. Publish SNAPSHOT to RepoA.
3. Try to use the dependency.
4. Publish SNAPSHOT to RepoB.
5. Try to use the dependency.
## problem

You get the SNAPSHOT from RepoA.
## expectation

You always get the fresh one.
## Notes
- https://github.com/sbt/sbt/tree/fix/issue-321-snapshot-multiple-repo
- [sbt-dirty-money](https://github.com/sbt/sbt-dirty-money)

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes #1514, #321

Adds `lastestSnapshots` flag to `updateOptions`, which controls the behavior of the chained resolver. Up until 0.13.6, sbt was picking the first `-SNAPSHOT` revision it found along the chain. When  `lastestSnapshots` is enabled (default: `true`), it will look into all resolvers on the chain, and compare them using the publish date.
The tradeoff is probably a longer resolution time if you have many remote repositories on the build or you live away from the severs. So here's how to disable it:

``` scala
updateOptions := updateOptions.value.withLatestSnapshots(false)
```

Ivy by default uses latest-revision as the latest strategy. This strategy I don't think takes in account for the possibility that a changing revision may exist in multiple repositories/resolvers with having identical version number like 0.1.0-SNAPSHOT.
The implementation is a bit hacky, but I think it attacks the core of this problem.

### CODE REVIEW

File: ivy/src/main/scala/sbt/Ivy.scala [hunk 1]
```diff
@@ -285,8 +294,187 @@ private object IvySbt {
...
+                Message.verbose(getName + ": Checking cache for: " + dd)
+                Option(findModuleInCache(dd, data, true)) map { mr =>
+                  Message.verbose(getName + ": module revision found in cache: " + mr.getId)
+                  forcedRevision(mr)
+                }
+              }
+            var temp: Option[ResolvedModuleRevision] =
+              if (useLatest) None
+              else resolvedOrCached
+            val resolvers = getResolvers.toArray.toVector collect { case x: DependencyResolver => x }
```
Comment (jsuereth) [inline ivy/src/main/scala/sbt/Ivy.scala:326]: Why the filter?  What else could be in there?

File: ivy/src/main/scala/sbt/Ivy.scala [hunk 2]
```diff
@@ -285,8 +294,187 @@ private object IvySbt {
...
+                }
+              case None if errors.size > 1 =>
+                val err = (errors.toList map { IvyStringUtils.getErrorMessage }).mkString("\n\t", "\n\t", "\n")
+                throw new RuntimeException(s"several problems occurred while resolving $dd:$err")
+              case _ =>
+                if (resolved == mrOpt) resolved.orNull
+                else (mrOpt map { resolvedRevision }).orNull
+            }
+          }
+        // Ivy seem to not want to use the module descriptor found at the latest resolver
```
Comment (jsuereth) [inline ivy/src/main/scala/sbt/Ivy.scala:396]: That' the fun stale cache thing I found IIRC.  It tries to keep MDs in memory.

File: ivy/src/main/scala/sbt/UpdateOptions.scala [hunk 3]
```diff
@@ -9,19 +9,26 @@ import java.io.File
  *
  * See also UpdateConfiguration in IvyActions.scala.
  */
-final class UpdateOptions(
+final class UpdateOptions private[sbt] (
+    /** If set to true, check all resolvers for snapshots. */
+    val latestSnapshots: Boolean,
```
Comment (jsuereth) [inline ivy/src/main/scala/sbt/UpdateOptions.scala:14]: Maybe this should instead be an option of which chain resolver you pick?

I guess that means users wouldn't be able to disable it...

**Comment by jsuereth** [comment]
LGTM!!!!

File: ivy/src/main/scala/sbt/Ivy.scala (hunk 1 above)
Comment (eed3si9n) [inline ivy/src/main/scala/sbt/Ivy.scala:326]: Ivy doesn't use Java 5 generics, so we get like `Collection` all the time.

**Pull request merged**